hpke/algs.go

```diff
@@ -0,0 +1,151 @@
+package hpke
+
+import (
+	"crypto"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/elliptic"
+	"errors"
+
+	"github.com/cloudflare/circl/dh/x25519"
+	"github.com/cloudflare/circl/dh/x448"
+	"github.com/cloudflare/circl/ecc/p384"
+	"github.com/cloudflare/circl/kem"
+	"golang.org/x/crypto/chacha20poly1305"
+)
+
+type KemID uint16
+
+const (
+	// DHKemP256HkdfSha256 is a KEM using P256 curve with HKDF based on SHA-256.
+	DHKemP256HkdfSha256 KemID = 0x10
+	// DHKemP384HkdfSha384 is a KEM using P384 curve with HKDF based on SHA-384.
+	DHKemP384HkdfSha384 KemID = 0x11
+	// DHKemP521HkdfSha512 is a KEM using P521 curve with HKDF based on SHA-512.
+	DHKemP521HkdfSha512 KemID = 0x12
+	// DHKemX25519HkdfSha256 is a KEM using X25519 Diffie-Hellman function with
+	// HKDF based on SHA-256.
+	DHKemX25519HkdfSha256 KemID = 0x20
+	// DHKemX448HkdfSha512 is a KEM using X448 Diffie-Hellman function with HKDF
+	// based on SHA-512.
+	DHKemX448HkdfSha512 KemID = 0x21
+)
+
+func (k KemID) Scheme() kem.AuthScheme {
+	switch k {
+	case DHKemP256HkdfSha256:
+		return dhkemP256HkdfSha256
+	case DHKemP384HkdfSha384:
+		return dhkemP384HkdfSha384
+	case DHKemP521HkdfSha512:
+		return dhkemP521HkdfSha512
+	case DHKemX25519HkdfSha256:
+		return dhkemX25519HkdfSha256
+	case DHKemX448HkdfSha512:
+		return dhkemX448HkdfSha512
+	default:
+		return nil
+	}
+}
+
+type KdfID uint16
+
+const (
+	// HKDF using SHA-256 hash function.
+	HkdfSha256 KdfID = 0x01
+	// HKDF using SHA-384 hash function.
+	HkdfSha384 KdfID = 0x02
+	// HKDF using SHA-512 hash function.
+	HkdfSha512 KdfID = 0x03
+)
+
+func (k KdfID) Hash() crypto.Hash {
+	switch k {
+	case HkdfSha256:
+		return crypto.SHA256
+	case HkdfSha384:
+		return crypto.SHA384
+	case HkdfSha512:
+		return crypto.SHA512
+	default:
+		return crypto.Hash(0)
+	}
+}
+
+type AeadID uint16
+
+const (
+	// AES-128 block cipher in Galois Counter Mode (GCM).
+	AeadAES128GCM AeadID = 0x01
```

Comment:
It's odd that abbreviations are capitalized here but not elsewhere. E.g., "HkdfSha256" above and "AES128GCM" here. I suggest changing to "Aes128Gcm" for consistency. Here and below.